docs/asdf/developer_overview.rst

```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
+same trick that pyyaml does with generators. Although pyyaml creates
+references between basic python objects, these references must be
+converted to references between ASDF objects, and doing so requires 
+a similar mechanism for building the ASDF objects. The 
+``_TreeModificationContext`` object (hereafter context object)
+holds the incomplete generators a similar way the pyyaml
+``construct_document`` function did. 
+
+There are differences though. It provides
+methods to indicate if nodes are pending (i.e., incomplete), and there
+is a special value ``PendingValue`` that is a signal that the node hasn't 
+been handled yet (e.g., it may be referencing something yet to be done).
+If ``PendingValue`` persists to the end, it indicates a failure to handle 
+circular references in the tag code. This approach was taken because 
+one of the earlier prototype implementations did something like this, 
+passing dict and list subclasses that would throw an exception if a
+``PendingValue`` element was accessed.  That would have been more friendly
+to extension developers, but it was discarded because it wasn't thougth
+it was worth turning all those high performance containers into slower
+asdf subclasses.  We may want to revisit this if we decide to implement
+a tree that tracks "dirty" nodes and only writes to disk those that
+have changed, since in that case we'll need custom container subclasses
+anyway.  We could also consider writing our own dict/list subclass in C
+so we could have our cake and eat it too.
+
+The ``walk_and_modify`` code handles the case where the tag code returns 
+a generator instead of a value. These generators are expected to be the
+similar kind of generator that pyyaml uses, but differing in that instead
+of returing an empty container object it will populate whatever elements
+it can completely (e.g, all non-mutable ones), and completing the 
+population of all the mutable memebers on the second iteration
+(which may, in turn, generate new generators for mutable elements 
+contained within). In handling these generators, the 
+first iteration also saves the generator in the context. When the
+top level of the context is reached (it handles nesting by indicating
+how many times it has been entered as a context), it starts "draining"
+the saved generators by doing the second iteration on them. Like 
+pyyaml, this second iteration may produce yet more generators that
+get saved, and thus keeps iterating on the saved generators until none
+are left.
+
+It is not possible to construct reference cycles in immutable
+objects within pure Python code, and thus the generators are only needed
+for mutable constructs (e.g., dicts and lists).
+
+Historical note: versions of the ASDF library prior to 2.6.0 required
+tag code when converting from a tagged object to a custom object to 
+call ``tagged_tree_to_custom_tree`` on any values of attributes that may be
+arbitrarily nested objects. That no longer is needed with the latest code
```

Comment:
Looking back at the old treeutil code, I'm not sure why this was necessary before.  The old code seems to be transforming the children first:

https://github.com/spacetelescope/asdf/blob/2.5.2/asdf/treeutil.py#L137

Maybe the astropy tag classes were calling tagged_tree_to_custom_tree needlessly?

Reply:
Should we say that?

Reply:
I'm content to just sweep that detail under the rug.